ironfish/src/memPool/priorityQueue.ts

```diff
@@ -0,0 +1,227 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+/**
+ * PriorityQueue keeps track of two data structures: a binary heap (_items) and a map
+ * of objects to their position in the heap (_positions). This map allows for `log(n)`
+ * removal from the heap instead of doing a linear scan of the heap to remove an item
+ */
+export class PriorityQueue<T> {
+  private _items: T[] = []
+  private _positions: { [key: string]: number } = {}
+  private _compare: (a: T, b: T) => boolean
+  hash: (item: T) => string
+
+  /**
+   * This data structure is a **Priority Queue** as well as a **Set**. Duplicate entries are
+   * not allowed. Entries are identified based on the hash function provided
+   *
+   * @param compare should return true if a is higher priority than b, otherwise false
+   * @param hash should return a unique hash of the item as a string. It should and should
+   * be a pure function and very fast (ideally just a field access) because it is called often
+   */
+  constructor(compare: (a: T, b: T) => boolean, hash: (item: T) => string) {
+    this._compare = compare
+    this.hash = hash
```

Comment:
Could hash be private as well? I'm curious why hash is public and _compare is private